tools/real_time_graphing.py

```diff
@@ -0,0 +1,302 @@
+
+
+class RealTimeGraph:
+    """
+    Tool to graph data as it comes in in real time.
+
+    Version: python 2.7 / 3.6
+
+    Requirements: Numpy, Matplotlib
+
+    Parameters
+    ----------
+    get_data: func
+        Function data reader will call to get data
+
+    pan_width: int
+        Time in seconds to display previous data
+    """
+
+    log_level = logging.INFO
+
+    max_rows = 3  # Rows of subplots per column
+
+    data_freq_warning = .5  # If time values are this far apart warn the user
+
+    pan_width = 10  # Seconds of previous data to show
+
+    working_filename = 'real_time_graphing.py'  # The name of this file
+
+    relative_config_path = "real_time_graph/config.xml"  # Path from this file to config file
+
+    def __init__(self, get_data, **kwargs):
+        printer = logging.getLogger()
+
+        if not printer.handlers:
+            printer.setLevel(RealTimeGraph.log_level)
+            handler = logging.StreamHandler()
+            handler.setLevel(RealTimeGraph.log_level)
+            printer.addHandler(handler)
+
+        if not get_data: logging.critical("RTG: No data pull function!")
+        self.get_data = get_data
+        self.pan_width = abs(kwargs['pan_width']) if 'pan_width' in kwargs.keys() else RealTimeGraph.pan_width
+
+        # Stored which data items we are interested in
+        self.tracked_data = []
+
+        self.plot_count = 0
+        self.data_count = 0
+
+        # Stores times corresponding to each data index
+        self.times = []
+
+        self.check_time = self.start_time = time()
+
+        # Initializes figure for real_time_graph
+        plt.rcParams['toolbar'] = 'None'  # Disable matplot toolbar
+
+        self.fig = plt.figure(figsize=(8, 6))
+        self.fig.canvas.set_window_title('Real Time Graphing')
+
+        self.fig.subplots_adjust(hspace=1, wspace=0.75)  # Avoid subplot overlap
+
+        # Threading
+        self.sleep_time = kwargs['sleep_time'] if 'sleep_time' in kwargs.keys() else 1e-1
+
+        self.thread_stop = kwargs['thread_stop'] if 'thread_stop' in kwargs.keys() else threading.Event()
+
+        self.thread_queue = Queue()
+
+    def run(self):
+        self.parse_rtg_config()
+
+        self.ani = animation.FuncAnimation(self.fig, self.plot_data, blit=False, interval=20, repeat=False)
+
+        threads = {
+            'reader': threading.Thread(target=self.read_data, args=(self.thread_queue,)),
+            'processor': threading.Thread(target=self.process_data, args=(self.thread_queue,))
+        }
+
+        for thread in threads.values():
+            thread.start()
+
+        plt.show()
+
+        #
+        # Cleanup
+        self.thread_stop.set()
+
+        for thread in threads.values():
+            thread.join()
+
+    @property
+    def config_filename(self):
+        """
+        Returns the filename of config file.
+        """
+
+        config_path = __file__.split(RealTimeGraph.working_filename)[0]  # Get this files location
```

Comment:
no. this is what os.path funcions are for
you can use the following code instead
```suggestion
        config_path = os.path.dirname(__file__)  # Get this files location
```